pyanaconda/modules/payload/base/handler_base.py

```diff
@@ -19,15 +19,82 @@
 from pyanaconda.modules.common.base import KickstartBaseModule
 
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
 
 class PayloadHandlerBase(KickstartBaseModule, metaclass=ABCMeta):
     """Base class for all the payload handler modules.
 
     This will contain all API specific to payload handlers which will be called
     by the base payload module.
     """
+    def __init__(self):
+        super().__init__()
+        self._sources = []
+        self.sources_changed = Signal()
+
+    @property
+    @abstractmethod
+    def supported_source_types(self):
+        """Get list of supported source types.
+
+        :return: list of supported source types
+        :rtype: [values from payload.base.constants.SourceType]
+        """
+        pass
+
+    @property
+    def sources(self):
+        """Get list of sources attached to this payload handler.
+
+        :return: list of source objects attached to this handler
+        :rtype: [instance of PayloadSourceBase class]
+        """
+        return self._sources
+
+    def set_sources(self, sources):
+        """Set a new list of sources to this payload handler.
+
+        Before setting the sources, please make sure the sources are not initialized otherwise
+        the SourceSetupError exception will be raised. Payload have to cleanup after itself.
+
+        ..NOTE:
+        The SourceSetupError is a reasonable effort to solve the race condition. However,
+        there is still a possibility that the task to initialize sources (`SetupSourcesWithTask()`)
+        was created with the old list but not run yet. In that case this check will not work and
+        the initialization task will run with the old list.
+
+        :param sources: set a new sources
+        :type sources: instance of pyanaconda.modules.payload.base.source_base.PayloadSourceBase
+        :raise: IncompatibleSourceError when source is not a supported type
+                SourceSetupError when attached sources are initialized
+        """
+        # TODO: Add test for this when there will be public API
+        for source in sources:
+            if source.type not in self.supported_source_types:
+                raise IncompatibleSourceError("Source type {} is not supported by this handler."
+                                              .format(source.type))
+
+        if any([source.is_ready for source in self.sources]):
+            raise SourceSetupError("Can't change list of sources if there is at least one source "
+                                   "initialized! Please tear down the sources first.")
+
+        self._sources = sources
+        log.debug("New sources %s was added.", sources)
+        self.sources_changed.emit()
+
+    def has_source(self):
+        """Check if any source is set.
+
+        :return: True if source object is set
+        :rtype: bool
+        """
+        return True if self.sources else False
```

Comment:
You should return just `bool(self.sources)`.

Reply:
Fixed.